### PR TITLE
Merge of the Chuangmi Plug V1, V2, V3 and M1

### PR DESCRIPTION
**Pull request by syssi**
@rytilahti Could you provide a recommendation for the "device model enum" and the discovery device map?

### CODE REVIEW

File: miio/device.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ class Device:
     This class should not be initialized directly but a device-specific class inheriting
     it should be used instead of it."""
     def __init__(self, ip: str = None, token: str = None,
-                 start_id: int=0, debug: int=0, lazy_discover: bool=True) -> None:
+                 start_id: int=0, debug: int=0, lazy_discover: bool=True, model: str=None) -> None:
```
Comment (rytilahti) [inline miio/device.py]: Is it truly necessary to introduce this also on the parent level, or would it be fine to have it just in the subclass?

File: miio/discovery.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
     "roborock-vacuum-s5": Vacuum,
     "chuangmi-plug-m1": Plug,
     "chuangmi-plug-v2": Plug,
-    "chuangmi-plug-v1": PlugV1,
-    "chuangmi-plug_": PlugV1,
-    "chuangmi-plug-v3": PlugV3,
+    "chuangmi-plug-v1": ChuangmiPlug,
+    "chuangmi-plug_": ChuangmiPlug,
+    "chuangmi-plug-v3": ChuangmiPlug,
```
Comment (rytilahti) [inline miio/discovery.py]: For passing a parameter here you can use `functools.partial`, so something like:
```
partial(ChuangmiPlug, model=MyModel)
```
can be used in this case.

I'm wondering whether that should be an enum, or simply an integer `model_version`, which would default to either of them? We could also derive the model (if not given) per default from the status information (to adapt the setters accordingly), as it's just one extra attribute & different naming?

File: miio/__init__.py [hunk 3]
```diff
@@ -1,11 +1,10 @@
 # flake8: noqa
 from miio.protocol import Message, Utils
 from miio.vacuumcontainers import (VacuumStatus, ConsumableStatus, DNDStatus,
-                                   CleaningDetails, CleaningSummary, Timer)
+                                   CleaningDetails, CleaningSummary, Timer, )
```
Comment (rytilahti) [inline miio/__init__.py]: Is this a necessary change?

Comment (syssi) [reply miio/__init__.py]: No. :-)

File: miio/discovery.py (hunk 2 above)
Comment (syssi) [reply miio/discovery.py]: I will care about tomorrow.

Comment (syssi) [reply miio/discovery.py]: The `check_and_create_device` method doesn't enter the `inspect.isclass(v)` branch anymore if I use a partial. I'm unsure about a good implementation.

Comment (rytilahti) [reply miio/discovery.py]: Indeed.. Hmm, maybe adding a check if it's partial and extracting the class out from it?
```
>>> from functools import partial
>>> class A:
>>>     pass

>>> partial(A, foo=1).func
<class '__main__.A'>
```

**Comment by rytilahti** [comment]
One open question is whether we want to keep backwards compatibility with the old class naming? If yes, that could be simply done by creating small wrapper classes located inside `chuangmi_plug`, which are also exported (and yielding a warning about deprecation).

**Comment by syssi** [comment]
Could you review the changes? I don't know how to improve the tests. The constructor of the old classes should be called and the base class must be mocked somehow.

File: miio/__init__.py [hunk 4]
```diff
@@ -3,9 +3,10 @@
 from miio.vacuumcontainers import (VacuumStatus, ConsumableStatus, DNDStatus,
                                    CleaningDetails, CleaningSummary, Timer)
 from miio.vacuum import Vacuum, VacuumException
-from miio.plug import Plug
-from miio.plug_v1 import PlugV1
-from miio.plug_v3 import PlugV3
+from miio.chuangmi_plug import Plug
+from miio.chuangmi_plug import PlugV1
+from miio.chuangmi_plug import PlugV3
+from miio.chuangmi_plug import ChuangmiPlug
```
Comment (rytilahti) [inline miio/__init__.py]: `from miio.chuangmi_plug import (Plug, PlugV1, PlugV3, ChuangiPlug)` would be nicer here.

File: miio/chuangmi_plug.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        warnings.warn(
+            "This device class is deprecated. Please use the ChuangmiPlug "
+            "in future and select a model by parameter 'model'.",
+            DeprecationWarning, stacklevel=2)
+
+        super().__init__(ip, token, start_id, debug, lazy_discover,
+                         model=model)
+
+
+class Plug(ClassDeprecatedPlug):
```
Comment (rytilahti) [inline miio/chuangmi_plug.py]: I created PR #276 to move the deprecated decorator out from vacuumcontainers, it should be used to decorate these classes to inform anyone trying to initialize these classes.

**Comment by syssi** [comment]
Done :-)

**Comment by rytilahti** [comment]
Great, thanks! :100:

**Pull request merged**